Tool Calling and Parallel Function Calling

Archimagus opened the pull request:
This PR adds support for OpenAI's new tool_calls feature which by its nature includes parallel function calling. 

resolves #729
resolves #732 

Currently still WIP, Looking for feedback/collaboration.  Especially with implementation for frameworks other than svelte.

Archimagus commented:
I'm not sure why the tests are failing. It seems unrelated to my work.  I'll continue to investigate, but if anyone knows what's up please chime in.

avianion commented:
Looks good. How long do you think this will take to get productionized?

Archimagus commented:
The functionality is fully there in the svelte version of useChat, as well as all the generic library parts (Types, OpenAIStream, etc.) It just needs implementation in the other libraries, Next, etc.  
And then of course it needs reviewed to see if anyone wants to see things done differently.
If I do the other implementations it'll likely take a few day possibly through next weekend.  If someone else wants to contribute to this PR to do the other platforms that will obviously speed things along.

MaxLeiter commented:
Thanks so much, @Archimagus! I like the direction of this 🎉 

I haven't looked too closely, and feel free to address this further along, but it'd be great if the markdown file could be converted into tests.

johannbuscail commented:
Hey, when do you think this will this be published on npm ?

Archimagus commented:
> Hey, when do you think this will this be published on npm ?

I've been slowly chipping away at it in the evenings.  I'm hoping I'll have more time to wrap it up this weekend.  Then, however long it takes to get it approved.

Archimagus commented:
OK, I think all the functionality is in for the generic code, and in useChat for svelte.  I think this is enough to get this merged once reviewed so people can start using it on the server side, and in svelte.  I'm not as familiar with the other language specific implementations, so I'd appreciate it if someone that uses those platforms would extend it for them.

lgrammel commented:
@Archimagus adding tool support would be awesome! this PR looks mostly good. A few thoughts from my end:

1. a quick heads up: this PR will cause a conflict with the upcoming PR that adds data stream support to Svelte/`useChat`
2. you changed the `next/openai` example, but the actual change is in the Svelte hook. Wouldn't the Svelte example need to be changed instead?
3. `stream-sample.md` could be deleted now that it seems to be included in the tests
4. (important) you seem to be using the regular function calls, not the data streams. Going forward useChat will likely be using only data streams (once it's implemented in for all frameworks)

Archimagus commented:
I haven't actually updated any of the examples yet,  just the couple lines in the next examples that were causing a type error.  I can update the svelte example to use the new system though. 

I'm not sure about the data streams, I just extended what was already there to handle the new tool calls API.

lgrammel commented:
> I haven't actually updated any of the examples yet, just the couple lines in the next examples that were causing a type error. I can update the svelte example to use the new system though.
> 
> I'm not sure about the data streams, I just extended what was already there to handle the new tool calls API.

Data streams are what's used by `experimental_StreamData`. It sends different events as lines. You can find the different line types here: https://github.com/vercel/ai/blob/main/packages/core/shared/stream-parts.ts

There is a lot of churn in that area right now tho, bc we are adding data streams support to every `useChat` and `useCompletion` hook.

Archimagus commented:
OK, @lgrammel I think I have ported it to use the new stuff.  But I am not sure that I got everything.  Also, when trying to use it, tool calls are not getting back to the client even if ignored by the server.  I don't suppose you could take a look and see what I'm missing, or if I've completely messed it up 🤨

lgrammel commented on packages/core/shared/stream-parts.ts:
```diff
@@ -48,9 +48,50 @@ const functionCallStreamPart: StreamPart<
-
-const dataStreamPart: StreamPart<'2', 'data', Array<JSONValue>> = {
+const toolCallStreamPart: StreamPart<
+  '2',
```
@Archimagus the tests are currently broken for me when I run `pnpm test` inside `packages/core` on your branch. Reverting the numbering and using a new number for tool calls would probably fix most of the tests. Please let me know when the tests pass and I'll take another look.

Archimagus replied:
I think the tests were also broken on main when I merged last,  I'll check again.  And sure I can use a new number for tools.

Archimagus replied:
So, it all seems to be there, the the UI tests all hang forever.  I am looking, but as of right now I'm kind of stumped.  If I skip all the UI tests then everything is fine, but enabling any of them and it just hangs forever.  I can't even get a console log to print from anywhere in a test.

lgrammel replied:
I just checked out your PR and all tests (incl. UI) are passing for me locally (at least with `pnpm test` inside `packages/core`).

Archimagus replied:
Well that's infuriating.  The UI tests never end on my machine, I see all the tests pass for both edge and node, and then it tries to start the UI and it just spins.  Same thing on the github, the tests are failing because they time out after 30 minutes.  (They don't even obey timeouts specified on the tests) It's like it can't initialize something.  I've tried pnpm install, pnpm rebuild,  deleting my node_modules folder, so far I can't figure out what is causing them to hang.

lgrammel replied:
Btw, I'm working on switching the tests over to Vite, 🤞🏻 that this solves a bunch of issues if it works.

Archimagus replied:
I'll check my Node version when I get home.  But that still doesn't explain why it fails on the github PR.  Either way,  yeah Vite test would be amazing.

lgrammel replied:
heads-up: the vitest PR has landed. maybe that helps with your testing issues

lgrammel commented on packages/core/streams/openai-stream.test.tsx:
```diff
@@ -365,6 +366,103 @@ describe('OpenAIStream', () => {
+      const chunks = await client.readAll();
+
+      expect(chunks).toEqual([
+        '2:[{"function_name":"get_date_time"},{"function_name":"open_webpage"}]\n',
+        '0:"experimental_onToolCall-return-value"\n',
```
I'm a bit confused here. Wouldn't there be a `6:some-tool-call-information` message that's forwarded to the client when there is a tool call?

There is a similar test for functions:
```ts
    it('should send function response and data when onFunctionCall is defined, returns undefined, and data is added', async () => {
      const data = new experimental_StreamData();

      const stream = OpenAIStream(
        await fetch(server.api + '/mock-func-call', {
          headers: {
            'x-mock-service': 'openai',
            'x-mock-type': 'func_call',
          },
        }),
        {
          onFinal() {
            data.close();
          },
          async experimental_onFunctionCall({ name }) {
            data.append({ fn: name });

            // no response
          },
          experimental_streamData: true,
        },
      );

      const response = new StreamingTextResponse(stream, {}, data);

      const client = createClient(response);
      const chunks = await client.readAll();

      expect(chunks).toEqual([
        '2:[{"fn":"get_current_weather"}]\n',
        '1:{"function_call":{"name":"get_current_weather","arguments":"{\\n\\"location\\": \\"Charlottesville, Virginia\\",\\n\\"format\\": \\"celsius\\"\\n}"}}\n',
      ]);
    });
    ```
    
Ideally having a test that exactly mimics that for tools would be good, and then additional test for additional behavior.

Archimagus replied:
I thought so too.  But and I had it like that, but it was always returning 2 instead of 6.
It was too the point where I was thinking (Oh, maybe that number in these tests is the chunk number?)
I can certainly add a more basic tool call test that is like this one, I didn't see that one for function calling.

lgrammel replied:
Hm that's weird, it points to a potential issue. Do you see messages of type 6 delivered in the Chrome network monitor when you run your Svelte example?

Archimagus replied:
I figured this one out.  It's because I am returning a string from the function call, which causes it (by design) to send that string back to the client.  hence the `'0:"experimental_onToolCall-return-value"\n',`
and then the other 2: is from the `data` passed in .

I implemented another simpler test where the "server" doesn't handle the call, and so you get the new message back to the client with the 6: number.

lgrammel commented:
Looks great and is getting close. For the svelte example, could you introduce a new tool example and leave the existing functions example as is? That would make it easy for me to test it out and preserve the old example for folks who are using function calls.

lgrammel commented:
> Well, I merged your latest. And at least now I can iterate on failing faster 😆 I was running node v20.3.1 I tried updating via `nvm install node Now using node v21.2.0 (npm v10.2.3)
> 
> Same issue, any of the UI tests just hang. Installed the Vite test runner for vscode and ran one of the tests through there and at least it gave me an error. `document is not defined` on the call to `render(<TestComponent />);`
> 
> Tried adding a dev dependency and including `// @vitest-environment happy-dom` at the top of the file, that just got be back to hanging with no warning or error, both when running pnpm test:ui or though the vscode test running.
> 
> Only other thing that I can think of that might be relevant is I'm developing in WSL2. Which I would suspect might be an issue, except the tests also fail on github.

There is an OOM issue in the CI tests:

![image](https://github.com/vercel/ai/assets/205036/ea2ba12f-4752-41cb-a2e1-47e1f41d76c4)

Since it only shows up here, it could be related to the PR.

Archimagus commented:
@lgrammel GOT IT!  Found the infinite loop that was hanging those tests.  I think all is well now.
Also I separated out the other parts as #820

lgrammel commented on packages/core/shared/process-chat-stream.ts:
```diff
@@ -56,32 +86,129 @@ export async function processChatStream({
+      }
     }
   }
+  if (count >= maxIterations) {
+    console.error(
+      'Max iterations reached for processChatStream probably an infinite ly an infinite loop.',
+    );
+  }
```
This got me thinking, why would this happen for tools but not for functions? It could be many things:

1. missing test for function
2. something different in the tools test
3. something different between functions / tools

Can the infinite loop (that's not infinite any more) happen outside test situations? why does it happen?

lgrammel commented:
> @lgrammel GOT IT! Found the infinite loop that was hanging those tests. I think all is well now. Also I separated out the other parts as #820

Awesome, the PR looks pretty good. The only issues I see is understanding the infinite loop and why it hit the tools tests, and thinking about react (bc experimental features get implemented in react 1st from what I understand).

CC @MaxLeiter

Archimagus commented:
The bug was due to this code
```typescript
if (!hasFollowingResponse) {
          break;
        }
```
being inside the 
```typescript
      // no sense parsing messages if there are no callbacks to call
      if (experimental_onFunctionCall || experimental_onToolCall) {
```
block
There was nothing to break the loop when neither a function or tool call was defined.  I refactored that section to handle those cases.  Then I added the safety valve, just in case, so if something like that happens again, then at least the tests will fail rather than hang forever.

lgrammel reviewed (approved):
Approved from my end for the current state. Would be good to get another review. Sadly this is pretty cross-cutting, and there is a lot of churn right now that affects this PR.

Archimagus commented:
I merged and fixed the conflicts.  Can we get this merged soon?  Having to keep dealing with merge conflicts is getting a bit stale.

Archimagus commented:
![SighOfReliefGIF](https://github.com/vercel/ai/assets/7319949/aed28518-b2a4-4290-a1d6-f4fd627fa3f1)

rajdtta commented:
@Archimagus apologies for the tag, but do you know whether this supports parallel tool calls where some tools are on the server, and other tools are on the client? 

From my initial experimentation, I can only get a response to work properly if all the tools being called are on the server or on the client.

Archimagus commented:
@rajdtta No worries.  It's supposed to, but I'm not certain I've tested that fully.  I'll take a look at it when I get a bit of time.

rajdtta commented:
@Archimagus just wanted to follow up and see if you've found anything conclusive about this

Archimagus commented:
@rajdtta Sorry, I've been super slammed at my day job and haven't had time to get back to this.  

If you wanted to look into it, you could start at
https://github.com/vercel/ai/blob/e25f3cab30036034b107b28dfb9896afa20d5dcb/packages/core/streams/openai-stream.ts#L664

Here we are checking if the tool calls were handled on the server.  I "shouldn't" be to difficult to track which tools were handled, and then filter on that to see if we need to pass any of them on to the client.

rajdtta commented:
> @rajdtta Sorry, I've been super slammed at my day job and haven't had time to get back to this.
> 
> If you wanted to look into it, you could start at
> 
> https://github.com/vercel/ai/blob/e25f3cab30036034b107b28dfb9896afa20d5dcb/packages/core/streams/openai-stream.ts#L664
> 
> Here we are checking if the tool calls were handled on the server. I "shouldn't" be to difficult to track which tools were handled, and then filter on that to see if we need to pass any of them on to the client.

No worries :)

Thanks for the lead, I'll spend some time today looking into it

rajdtta commented:
> #1123
> 
> I've added a test case and code that "should" work to allow you to split handling tools across the server and client. So far there is only one test, and I haven't had a chance to test it in a real chat scenario. Feel free to take this and run with it, otherwise I will try and work on it more as time permits. But it won't likely be super soon.

Apologies for the late response, somehow missed the email notification for this. I'll play around with this sometime this week (or over the weekend outside of work).

Archimagus commented:
> Where is the best place to find documentation for this?

For the general tools use. I think the best would be to search for [chat-with-tools](https://github.com/search?q=repo%3Avercel%2Fai%20path%3Achat-with-tools&type=code) And use one of the examples as documentation.

frmsaul commented:
I see. Does anyone know if there is a way to return the function call decision to the? In the example, the model in the backend has access to the function call and function call return stuff but the frontend doesn't have access to it. Basically i'd like the frontend to know what function have been called and what is it's return value.

Line 97-104 of this [file](https://github.com/vercel/ai/blob/5f4115bdbba7a7fcaae15879b421a41b20620c2c/examples/next-openai/app/api/chat-with-tools/route.ts#L97)

Archimagus commented:
You can return a string from the function call on the server and that will get sent to the client as a standard message.  So you could do something like 
```javascript
      return JSON.stringify({function: toolCall.func.name, result: 'Handled'});
```
And then look for json formatted messages in the client.

If you want to return a function response to the API, (like the current weather) for the model to interpret and also tell the client that the function was handled, you might be able to get clever by adding some sort of `modelGotFunctionReturn` function on the client side where you tell the model to tell you about functions that were handled.  Not sure if that would work, but might be worth a try.

If none of that works, you could look into
https://github.com/vercel/ai/blob/5f4115bdbba7a7fcaae15879b421a41b20620c2c/packages/core/streams/openai-stream.ts#L673
and see if there is some changes you could make to add the functionality you are looking for.

I've been away from this code for a bit so I don't have too much more to offer than that.  Although, there is a draft PR up with some changes to that code that you could also look into for handing some calls on the server and others on the client.
https://github.com/vercel/ai/pull/1123

frmsaul commented:
Cool! I'll give it a shot.

frmsaul commented:
It works! Thank you so much

Archimagus commented:
>It works! Thank you so much

What did you try that worked?  Just would like to know.

frmsaul commented:
Returning the serialized json and then looking for jsons in the frontend
